packages/react/src/__tests__/state.test.tsx

```diff
@@ -0,0 +1,273 @@
+          {
+            $: 'component',
+            key: 'header',
+            component: 'View',
+            props: {
+              disabled: isDisabled,
+            },
+          },
+          {
+            $: 'component',
+            key: 'button',
+            component: 'View',
+            props: {
+              ['data-testid']: 'button',
+              onClick: event(setStateAction(isDisabled, toggle)),
+            },
+          },
+        ]
+      },
+    }),
+    sendEvent: jest.fn(),
+  }
+  const component = render(<Template components={BUILT_IN_COMPONENTS} dataSource={dataSource} />)
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div />
+      <div
+        data-testid="button"
+      />
+    </DocumentFragment>
+  `)
+  await act(async () => {
+    fireEvent.click(component.getByTestId('button'))
+  })
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div
+        disabled=""
+      />
+      <div
+        data-testid="button"
+      />
+    </DocumentFragment>
+  `)
+})
+
+it('should increment state', async () => {
+  const counter = state(1)
+  const dataSource: DataSource = {
+    get: () => ({
+      subscribe: () => jest.fn(),
+      get() {
+        return [
+          {
+            $: 'component',
+            key: 'header',
+            component: 'View',
+            children: counter,
+          },
+          {
+            $: 'component',
+            key: 'button',
+            component: 'View',
+            props: {
+              ['data-testid']: 'button',
+              onClick: event(setStateAction(counter, increment(2))),
+            },
+          },
+        ]
+      },
+    }),
+    sendEvent: jest.fn(),
+  }
+  const component = render(<Template components={BUILT_IN_COMPONENTS} dataSource={dataSource} />)
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div>
+        1
+      </div>
+      <div
+        data-testid="button"
+      />
+    </DocumentFragment>
+  `)
+  await act(async () => {
+    fireEvent.click(component.getByTestId('button'))
+    fireEvent.click(component.getByTestId('button'))
+  })
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div>
+        5
+      </div>
+      <div
+        data-testid="button"
+      />
+    </DocumentFragment>
+  `)
+})
+
+it('should modify state after receiving response from the server', async () => {
+  const value = state('foo')
+  const dataSource: DataSource = {
+    get: () => ({
+      subscribe: jest.fn().mockReturnValue(jest.fn()),
+      get() {
+        return {
+          $: 'component',
+          component: 'View',
+          props: {
+            ['data-testid']: 'button',
+            onClick: event(jest.fn()),
+          },
+          children: value,
+        }
+      },
+    }),
+    sendEvent: jest.fn().mockReturnValue(response(null).action(setStateAction(value, 'bar'))),
```

Comment:
hmm I'll think about the `response()` API.. in any case it should maybe have a more specific label like `eventResponse` so that it is more clear what the purpose is

Reply:
#63 

Yeah. The API should change, because it's not as good now as the library is growing. I have this open issue and plan to tackle is as a next thing, after rewriting all kit components to be server-only.